File: src/db/queries/prices.ts
import { and, asc, desc, eq, gte, lte, sql } from "drizzle-orm";
import { db, schema } from "@/db";

const { priceCache } = schema;

export type PriceRow = typeof priceCache.$inferSelect;

/** Most recent cached close per symbol (correlated MAX(date) subquery). */
export function getLatestPrices(): PriceRow[] {
  return db
    .select()
    .from(priceCache)
    .where(
      sql`${priceCache.date} = (SELECT MAX(date) FROM price_cache p2 WHERE p2.symbol = ${priceCache.symbol})`,
    )
    .all();
}

/** Insert OR IGNORE a batch of (date, close) rows in one transaction. */
export function bulkInsertPrices(
  symbol: string,
  rows: [string, number][],
): number {
  if (rows.length === 0) return 0;
  return db.transaction((tx) => {
    let inserted = 0;
    for (const [date, close] of rows) {
      inserted += tx
        .insert(priceCache)
        .values({ symbol, date, closePrice: close })
        .onConflictDoNothing()
        .run().changes;
    }
    return inserted;
  });
}

export function getPriceOnOrBefore(symbol: string, date: string): number | null {
  const row = db
    .select({ c: priceCache.closePrice })
    .from(priceCache)
    .where(and(eq(priceCache.symbol, symbol), lte(priceCache.date, date)))
    .orderBy(desc(priceCache.date))
    .limit(1)
    .get();
  return row?.c ?? null;
}

export function countHistorySince(symbol: string, since: string): number {
  const row = db
    .select({ c: sql<number>`COUNT(*)` })
    .from(priceCache)
    .where(and(eq(priceCache.symbol, symbol), gte(priceCache.date, since)))
    .get();
  return row?.c ?? 0;
}

/** All (date, close) on or before `until`, ascending — for ticker charts. */
export function listHistory(
  symbol: string,
  until: string,
): { date: string; close: number }[] {
  return db
    .select({ date: priceCache.date, close: priceCache.closePrice })
    .from(priceCache)
    .where(and(eq(priceCache.symbol, symbol), lte(priceCache.date, until)))
    .orderBy(asc(priceCache.date))
    .all();
}

/** Upsert a single close (ON CONFLICT DO UPDATE) — used by daily/refresh. */
export function upsertPrice(symbol: string, date: string, close: number): void {
  db
    .insert(priceCache)
    .values({ symbol, date, closePrice: close })
    .onConflictDoUpdate({
      target: [priceCache.symbol, priceCache.date],
      set: { closePrice: close },
    })
    .run();
}

File: src/db/prices.rs
use chrono::NaiveDate;
use sqlx::{FromRow, SqlitePool};

#[derive(Debug, FromRow)]
pub struct PriceCache {
    pub symbol: String,
    pub date: NaiveDate,
    pub close_price: f64,
}

pub async fn get_latest_prices(pool: &SqlitePool) -> Result<Vec<PriceCache>, sqlx::Error> {
    sqlx::query_as::<_, PriceCache>(
        r#"
        SELECT symbol, date, close_price
        FROM price_cache
        WHERE date = (SELECT MAX(date) FROM price_cache p2 WHERE p2.symbol = price_cache.symbol)
        "#,
    )
    .fetch_all(pool)
    .await
}

pub async fn upsert_price(
    pool: &SqlitePool,
    symbol: &str,
    date: NaiveDate,
    close_price: f64,
) -> Result<(), sqlx::Error> {
    sqlx::query(
        r#"
        INSERT INTO price_cache (symbol, date, close_price)
        VALUES (?, ?, ?)
        ON CONFLICT(symbol, date) DO UPDATE SET close_price = excluded.close_price
        "#,
    )
    .bind(symbol)
    .bind(date)
    .bind(close_price)
    .execute(pool)
    .await?;
    Ok(())
}

pub async fn bulk_insert_prices(
    pool: &SqlitePool,
    symbol: &str,
    rows: &[(NaiveDate, f64)],
) -> Result<usize, sqlx::Error> {
    if rows.is_empty() {
        return Ok(0);
    }
    // One transaction for the whole batch — backfills can run hundreds of
    // rows; without this each INSERT pays its own fsync.
    let mut tx = pool.begin().await?;
    let mut inserted = 0usize;
    for (date, close) in rows {
        let res = sqlx::query(
            "INSERT OR IGNORE INTO price_cache (symbol, date, close_price) VALUES (?, ?, ?)",
        )
        .bind(symbol)
        .bind(date)
        .bind(close)
        .execute(&mut *tx)
        .await?;
        inserted += res.rows_affected() as usize;
    }
    tx.commit().await?;
    Ok(inserted)
}

pub async fn get_price_on_or_before(
    pool: &SqlitePool,
    symbol: &str,
    date: NaiveDate,
) -> Result<Option<f64>, sqlx::Error> {
    let row: Option<(f64,)> = sqlx::query_as(
        "SELECT close_price FROM price_cache WHERE symbol = ? AND date <= ? ORDER BY date DESC LIMIT 1"
    )
    .bind(symbol)
    .bind(date)
    .fetch_optional(pool)
    .await?;
    Ok(row.map(|r| r.0))
}

/// Count cached price rows for `symbol` whose date is on or after `since`.
/// Used to decide whether a fresh history backfill is worth a yfinance call.
pub async fn count_history_since(
    pool: &SqlitePool,
    symbol: &str,
    since: NaiveDate,
) -> Result<i64, sqlx::Error> {
    let row: (i64,) =
        sqlx::query_as("SELECT COUNT(*) FROM price_cache WHERE symbol = ? AND date >= ?")
            .bind(symbol)
            .bind(since)
            .fetch_one(pool)
            .await?;
    Ok(row.0)
}

/// Fetch all (date, close) pairs for `symbol` on or before `until`,
/// ordered ascending — used for charting a ticker's history.
pub async fn list_history(
    pool: &SqlitePool,
    symbol: &str,
    until: NaiveDate,
) -> Result<Vec<(NaiveDate, f64)>, sqlx::Error> {
    let rows: Vec<(NaiveDate, f64)> = sqlx::query_as(
        r#"
        SELECT date, close_price FROM price_cache
        WHERE symbol = ? AND date <= ?
        ORDER BY date ASC
        "#,
    )
    .bind(symbol)
    .bind(until)
    .fetch_all(pool)
    .await?;
    Ok(rows)
}
